components/core/tests/test-NetworkReader.cpp

```diff
@@ -188,3 +190,53 @@ TEST_CASE("network_reader_illegal_offset", "[NetworkReader]") {
     size_t pos{};
     REQUIRE((clp::ErrorCode_Failure == reader.try_get_pos(pos)));
 }
+
+TEST_CASE("network_reader_with_custom_headers", "[NetworkReader]") {
+    std::map<std::string, std::string> custom_headers = std::map<std::string, std::string>();
+    // We use httpbin (https://httpbin.org/) to test the custom headers. This request will return a
+    // JSON object that contains the custom headers. We check if the headers are in the response.
+    constexpr int kNumRegularTestHeaders = 10;
+    for (int i = 0; i < kNumRegularTestHeaders; i++) {
+        std::string key = fmt::format("Unit-Test-Key{}", i);
+        std::string value = fmt::format("Unit-Test-Value{}", i);
+        custom_headers[key] = value;
+    }
+    // The following three headers are determined by offset and disable_cache, which should not be
+    // overrided by custom headers.
+    custom_headers["Range"] = "bytes=100-";
+    custom_headers["Cache-Control"] = "no-cache";
+    custom_headers["Pragma"] = "no-cache";
+    // Some illegal custom header names and values, which should not be added into headers
+    custom_headers["A Space"] = "xx";
+    custom_headers[R"(A\nNewline)"] = "xx";
+    custom_headers["An@At"] = "xx";
+    custom_headers["-Start-with-Non-Alphanumeric"] = "xx";
+    custom_headers["Legal-Name1"] = R"(newline\n)";
+    custom_headers["Legal-Name2"] = R"(control-char\x01)";
+    clp::NetworkReader reader{
+            "https://httpbin.org/headers",
+            0,
+            false,
+            clp::CurlDownloadHandler::cDefaultOverallTimeout,
+            clp::CurlDownloadHandler::cDefaultConnectionTimeout,
+            clp::NetworkReader::cDefaultBufferPoolSize,
+            clp::NetworkReader::cDefaultBufferSize,
+            custom_headers
+    };
+    auto const actual{get_content(reader)};
+    std::string actual_string(actual.begin(), actual.end());
+    REQUIRE(assert_curl_error_code(CURLE_OK, reader));
+    for (int i = 0; i < kNumRegularTestHeaders; i++) {
+        std::string field = fmt::format("\"Unit-Test-Key{}\": \"Unit-Test-Value{}\"", i, i);
+        REQUIRE(std::string::npos != actual_string.find(field));
+    }
```

Comment:
Instead of search for substings, we can parse the response using `nlohmann::json`, so we can validate them as key-value pairs. Here's a simple example:
```C++
    auto const content{nlohmann::json::parse(get_content(reader))};
    auto const& headers{content.at(0).at("headers")};
    REQUIRE(assert_curl_error_code(CURLE_OK, reader));
    for (size_t i{0}; i < cNumRegularTestHeaders; ++i) {
        REQUIRE((
                headers.at(fmt::format("Unit-Test-Key{}", i)) == fmt::format("Unit-Test-Value{}", i)
        ));
    }
```

Reply:
fixed in next commit